internal/servicelog/ringbuffer.go

```diff
@@ -0,0 +1,369 @@
+func (rb *RingBuffer) WriteTo(writer io.Writer, start RingPos) (int64, error) {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	if start < rb.readIndex || start > rb.writeIndex {
+		return 0, ErrRange
+	}
+	if rb.writeClosed && start == rb.writeIndex {
+		return 0, io.EOF
+	}
+	copyLength := rb.writeIndex - start
+	if copyLength == 0 {
+		return 0, nil
+	}
+	end := rb.writeIndex
+	buffers := rb.buffers(start, end)
+	written := int64(0)
+	for _, buffer := range buffers {
+		if len(buffer) == 0 {
+			continue
+		}
+		n, err := writer.Write(buffer)
+		written += int64(n)
+		if err != nil {
+			return written, err
+		}
+	}
+	if rb.writeClosed && start+RingPos(written) == rb.writeIndex {
+		return written, io.EOF
+	}
+	return written, nil
+}
+
+// TailIterator returns an iterator from the tail of the buffer.
+func (rb *RingBuffer) TailIterator() Iterator {
+	rb.iteratorMutex.Lock()
+	defer rb.iteratorMutex.Unlock()
+	start, _ := rb.Positions()
+	iter := &iterator{
+		rb:         rb,
+		index:      start,
+		notifyChan: make(chan struct{}),
+		closeChan:  make(chan struct{}),
+	}
+	if rb.Closed() {
+		close(iter.closeChan)
+	}
+	rb.iteratorList = append(rb.iteratorList, iter)
+	return iter
+}
+
+// HeadIterator returns an iterator from the head of the buffer.
+func (rb *RingBuffer) HeadIterator(lines int) Iterator {
+	firstLine := rb.reverseLinePosition(lines)
+	rb.iteratorMutex.Lock()
+	defer rb.iteratorMutex.Unlock()
+	iter := &iterator{
+		rb:         rb,
+		index:      firstLine,
+		notifyChan: make(chan struct{}),
+		closeChan:  make(chan struct{}),
+	}
+	if rb.Closed() {
+		close(iter.closeChan)
+	}
+	rb.iteratorList = append(rb.iteratorList, iter)
+	return iter
+}
+
+func (rb *RingBuffer) reverseLinePosition(n int) RingPos {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	if n <= 0 {
+		return rb.writeIndex
+	}
+	buffers := rb.buffers(rb.readIndex, rb.writeIndex)
+	// a line is not complete until newline is written, so start negative.
+	lines := -1
+	firstLine := rb.writeIndex
+	last := byte(0)
+out:
+	for j := len(buffers) - 1; j >= 0; j-- {
+		buf := buffers[j]
+		for i := len(buf) - 1; i >= 0; i-- {
+			firstLine--
+			last = buf[i]
+			if last == byte('\n') {
+				lines++
+			}
+			if lines == n {
+				break out
+			}
+		}
+	}
+	if last == byte('\n') {
+		firstLine++
+	}
+	return firstLine
+}
+
+// Discard disposes of n bytes from the tail of the buffer making
+// them available to be used for subsequent writes.
+func (rb *RingBuffer) Discard(n int) error {
+	rb.rwlock.Lock()
+	defer rb.rwlock.Unlock()
+	return rb.discard(n)
+}
+
+func (rb *RingBuffer) discard(n int) error {
+	buffered := rb.buffered()
+	if n > buffered {
+		n = buffered
+	}
+	rb.readIndex = rb.readIndex + RingPos(n)
+	return nil
+}
+
+func (rb *RingBuffer) signalIterators() {
+	rb.iteratorMutex.RLock()
+	defer rb.iteratorMutex.RUnlock()
+	for _, iter := range rb.iteratorList {
+		select {
+		case iter.notifyChan <- struct{}{}:
+		default:
```

Comment:
Given that these channels are unbuffered, the `default:` here will be used rule whenever the other side is busy, which means if it is busy and we try to notify it that there's more content being written one instruction before it decides to wait, it's dead and won't ever recognize the fact that there's more data pending before the next notification. For this logic to be correct, we need to have at least 1 item of buffer in the channel.

On a separate note, I also tend to have my channels as `bool` instead of `struct{}`, because code looks nicer and on the consuming side the bool sometimes make logic neater copying the byte back and forth tends to be irrelevant in terms of space or time. Whatever you prefer in this case, though.

Reply:
Good catch, these were supposed to be buffered channels with a size of 1.

Happy to change to bool, I think struct{} channels only make true sense for cancellation.